driver-core/src/main/com/mongodb/connection/InternalStreamConnection.java

```diff
@@ -259,12 +263,37 @@ public void sendMessageAsync(final List<ByteBuf> byteBuffers, final int lastRequ
 
     @Override
     public void receiveMessageAsync(final int responseTo, final SingleResultCallback<ResponseBuffers> callback) {
-        notNull("open", stream, callback);
+        isTrue("stream is open", stream != null, callback);
+
         if (LOGGER.isTraceEnabled()) {
             LOGGER.trace(format("Queuing read message: %s. ([%s] %s)", responseTo, getId(), serverId));
         }
-        readQueue.put(responseTo, errorHandlingCallback(callback, LOGGER));
-        processPendingReads();
+
+        if (isClosed()) {
+            callback.onResult(null, new MongoSocketClosedException("Can not read from a closed socket", getServerAddress()));
+        }
+
+        readerLock.lock();
+
+        Response response = messages.get(responseTo);
+
+        try {
+            if (response == null) {
```

Comment:
There are two possibilities:

# my response has already been read: if so, no need to add myself to the read queue.  Just grab the callback and execute it (after I give up the lock)
# my response has not yet been read: I need to put myself on the read queue.